File: shared/kotlin/botfarm/misc/RandomConfig.kt
package botfarm.misc

import kotlin.math.roundToInt
import kotlin.random.Random

class RandomConfig private constructor(
   // joshr: These can't be private because they are serialized
   val fixed: Double? = null,
   val range: Range? = null
) {
   companion object {
      val zero = RandomConfig.fixed(0.0)
      val one = RandomConfig.fixed(1.0)

      fun fixed(value: Number) = RandomConfig(
         fixed = value.toDouble()
      )

      fun range(min: Number, max: Number) = RandomConfig(
         range = Range(
            min = min.toDouble(),
            max = max.toDouble()
         )
      )
   }

   class Range(
      val min: Double,
      val max: Double
   )

   fun rollDouble(): Double {
      if (this.fixed != null) {
         return this.fixed.toDouble()
      }

      if (this.range != null) {
         return Random.nextDouble(this.range.min, this.range.max)
      }

      throw Exception("Invalid RandomConfig")
   }

   fun rollInt(): Int {
      return this.rollDouble().roundToInt()
   }
}

File: client/src/game/RandomConfig.ts
class Range {
  constructor(
    readonly min: number,
    readonly max: number
  ) {
  }
}

export class RandomConfig {
  readonly fixed: number | null;
  readonly range: Range | null;

  private constructor(fixed: number | null = null, range: Range | null = null) {
    this.fixed = fixed;
    this.range = range;
  }

  static zero = new RandomConfig(0.0, null);
  static one = new RandomConfig(1.0, null);

  static fixed(value: number): RandomConfig {
    return new RandomConfig(value, null);
  }

  static range(min: number, max: number): RandomConfig {
    return new RandomConfig(null, new Range(min, max));
  }

  static rollDouble(self: RandomConfig): number {
    if (self.fixed !== null) {
      return self.fixed;
    }

    if (self.range !== null) {
      return Math.random() * (self.range.max - self.range.min) + self.range.min;
    }

    throw new Error("Invalid RandomConfig");
  }

  static rollInt(self: RandomConfig): number {
    return Math.round(this.rollDouble(self));
  }
}